scene/2d/navigation_agent_2d.cpp

```diff
@@ -687,102 +695,145 @@ void NavigationAgent2D::update_navigation() {
+		const ObjectID waypoint_owner_id = ObjectID(navigation_path_owners[navigation_path_index]);
+
+		// Get a reference to the owning object.
+		Object *owner = nullptr;
+		if (waypoint_owner_id.is_valid()) {
+			owner = ObjectDB::get_instance(waypoint_owner_id);
+		}
 
-			// Move to the next waypoint on the list
-			navigation_path_index += 1;
-
-			// Check to see if we've finished our route
-			if (navigation_path_index == navigation_path.size()) {
-				_check_distance_to_target();
-				navigation_path_index -= 1;
-				navigation_finished = true;
-				target_position_submitted = false;
-				if (avoidance_enabled) {
-					NavigationServer2D::get_singleton()->agent_set_position(agent, agent_parent->get_global_position());
-					NavigationServer2D::get_singleton()->agent_set_velocity(agent, Vector2(0.0, 0.0));
-					NavigationServer2D::get_singleton()->agent_set_velocity_forced(agent, Vector2(0.0, 0.0));
+		details[SNAME("owner")] = owner;
+
+		if (waypoint_type == NavigationPathQueryResult2D::PATH_SEGMENT_TYPE_LINK) {
+			const NavigationLink2D *navlink = Object::cast_to<NavigationLink2D>(owner);
+			if (navlink) {
+				Vector2 link_global_start_position = navlink->get_global_start_position();
+				Vector2 link_global_end_position = navlink->get_global_end_position();
+				if (waypoint.distance_to(link_global_start_position) < waypoint.distance_to(link_global_end_position)) {
+					details[SNAME("link_entry_position")] = link_global_start_position;
+					details[SNAME("link_exit_position")] = link_global_end_position;
+				} else {
+					details[SNAME("link_entry_position")] = link_global_end_position;
+					details[SNAME("link_exit_position")] = link_global_start_position;
 				}
-				emit_signal(SNAME("navigation_finished"));
-				break;
 			}
 		}
 	}
-}
 
-void NavigationAgent2D::_request_repath() {
-	navigation_result->reset();
-	target_reached = false;
-	navigation_finished = false;
-	update_frame_id = 0;
+	// Emit a signal for the waypoint
+	emit_signal(SNAME("waypoint_reached"), details);
+
+	// Emit a signal if we've reached a navigation link
+	if (waypoint_type == NavigationPathQueryResult2D::PATH_SEGMENT_TYPE_LINK) {
+		emit_signal(SNAME("link_reached"), details);
+	}
 }
 
-void NavigationAgent2D::_check_distance_to_target() {
-	if (!target_reached) {
-		if (distance_to_target() < target_desired_distance) {
-			target_reached = true;
-			emit_signal(SNAME("target_reached"));
-		}
+void NavigationAgent2D::_transition_to_navigation_finished() {
+	navigation_finished = true;
+	target_position_submitted = false;
+
+	if (avoidance_enabled) {
+		NavigationServer2D::get_singleton()->agent_set_position(agent, agent_parent->get_global_position());
+		NavigationServer2D::get_singleton()->agent_set_velocity(agent, Vector2(0.0, 0.0));
+		NavigationServer2D::get_singleton()->agent_set_velocity_forced(agent, Vector2(0.0, 0.0));
 	}
+
+	emit_signal(SNAME("navigation_finished"));
+}
```

Comment:
`_trigger_waypoint_reached` and `_transition_to_navigation_finished` only contain code that was previously in `update_navigation`. The code is unchanged.